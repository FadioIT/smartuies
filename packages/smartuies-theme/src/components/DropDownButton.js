import React from 'react';
import PropTypes from 'prop-types';
import { DropDownButton, themeComponent } from '@fadioit/smartuies';
import Button from './Button';

const renderButton = ({ buttonRef, onToggle, open, buttonLabel }) => (
  <Button ref={buttonRef} onClick={onToggle}>
    {open ? '▾' : '▸'} {buttonLabel}
  </Button>
);
renderButton.propTypes = {
  buttonRef: PropTypes.any,
  onToggle: PropTypes.func.isRequired,
  open: PropTypes.bool.isRequired,
  buttonLabel: PropTypes.node.isRequired,
};

export const theme = {
  renderButton,
};

export default themeComponent(DropDownButton, theme);
