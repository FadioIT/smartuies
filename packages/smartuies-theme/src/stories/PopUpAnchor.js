import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';
import { PopUpAnchor } from '@fadioit/smartuies';
import { Window } from '../utils/storyBookUtils';
import Button from '../components/Button';

const { positions } = PopUpAnchor;

storiesOf('PopUpAnchor', module)
  .addDecorator(withKnobs)
  .add('basic example', () => {
    const popUpHeightMatchesAnchorHeight = boolean(
      'Match anchor height',
      false,
    );
    const popUpWidthMatchesAnchorWidth = boolean('Match anchor width', false);
    const displayPopUp = boolean('displayPopUp', true);
    const popUpPosition = select(
      'popUpPosition',
      {
        [positions.LEFT]: 'Left',
        [positions.RIGHT]: 'Right',
        [positions.ABOVE]: 'Above',
        [positions.BELOW]: 'Below',
        [positions.TOP_LEFT]: 'Top left',
      },
      positions.BELOW,
    );

    return (
      <PopUpAnchor
        popUpHeightMatchesAnchorHeight={popUpHeightMatchesAnchorHeight}
        popUpWidthMatchesAnchorWidth={popUpWidthMatchesAnchorWidth}
        popUpPosition={popUpPosition}
        anchor={<Button>Anchor element</Button>}
        displayPopUp={displayPopUp}
      >
        <Window>PopUp</Window>
      </PopUpAnchor>
    );
  });
