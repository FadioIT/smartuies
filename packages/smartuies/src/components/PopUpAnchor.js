import React from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

const PopUpPosition = {
  LEFT: 'left',
  RIGHT: 'right',
  ABOVE: 'above',
  BELOW: 'below',
  TOP_LEFT: 'topLeft',
};

class PopUpAnchor extends React.Component {
  static propTypes = {
    popUpHeightMatchesAnchorHeight: PropTypes.bool,
    popUpWidthMatchesAnchorWidth: PropTypes.bool,
    popUpPosition: PropTypes.oneOf([
      'left',
      'right',
      'above',
      'below',
      'topLeft',
    ]),

    anchor: PropTypes.element.isRequired,
    children: PropTypes.node,
    displayPopUp: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    popUpHeightMatchesAnchorHeight: false,
    popUpWidthMatchesAnchorWidth: false,
    popUpPosition: PopUpPosition.BELOW,
  };

  constructor(props, context) {
    super(props, context);
    this.popUpNode = document.createElement('div');
    document.body.appendChild(this.popUpNode);
  }

  componentDidMount() {
    if (window) {
      window.addEventListener('resize', this.windowResizeHandler);
    }
    this.updatePopUp();
  }

  componentWillUnmount() {
    document.body.removeChild(this.popUpNode);
    if (window) {
      window.removeEventListener('resize', this.windowResizeHandler);
    }
  }

  componentDidUpdate() {
    this.updatePopUp();
  }

  updatePopUp() {
    const {
      popUpHeightMatchesAnchorHeight,
      popUpWidthMatchesAnchorWidth,
      popUpPosition,
      displayPopUp,
    } = this.props;

    const { popUpNode } = this;
    popUpNode.style.display = 'none';
    // eslint-disable-next-line react/no-find-dom-node
    const domNode = ReactDOM.findDOMNode(this);
    const domNodeNodeBounds = domNode.getBoundingClientRect();

    popUpNode.style.display = displayPopUp ? null : 'none';

    const {
      width: popupNodeWidth,
      height: popupNodeHeight,
    } = popUpNode.firstChild.getBoundingClientRect();

    this.popUpNode.style.height = popUpHeightMatchesAnchorHeight
      ? `${domNodeNodeBounds.height}px`
      : null;

    this.popUpNode.style.width = popUpWidthMatchesAnchorWidth
      ? `${domNodeNodeBounds.width}px`
      : null;

    const nodeBounds = {
      left: domNodeNodeBounds.left,
      top: domNodeNodeBounds.top,
    };

    switch (popUpPosition) {
      case PopUpPosition.BELOW:
        nodeBounds.top += domNodeNodeBounds.height;
        break;
      case PopUpPosition.ABOVE:
        nodeBounds.top -= popupNodeHeight;
        break;
      case PopUpPosition.RIGHT:
        nodeBounds.left = domNodeNodeBounds.right;
        nodeBounds.top -= popupNodeHeight / 2 - domNodeNodeBounds.height / 2;
        break;
      case PopUpPosition.LEFT:
        nodeBounds.left -= popupNodeWidth;
        nodeBounds.top -= popupNodeHeight / 2 - domNodeNodeBounds.height / 2;
        break;
      case PopUpPosition.TOP_LEFT:
      default:
        // already 0,0
        break;
    }

    popUpNode.style.position = 'absolute';
    popUpNode.style.top = `${nodeBounds.top}px`;
    popUpNode.style.left = `${nodeBounds.left}px`;
  }

  render() {
    const { children, anchor } = this.props;

    const popup = ReactDOM.createPortal(children, this.popUpNode);

    return (
      <>
        {anchor}
        {popup}
      </>
    );
  }

  windowResizeHandler = () => {
    this.forceUpdate();
  };
}

PopUpAnchor.positions = PopUpPosition;

export default PopUpAnchor;
