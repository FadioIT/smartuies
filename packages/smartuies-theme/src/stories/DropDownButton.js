import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Window } from '../utils/storyBookUtils';

import DropDownButton from '../components/DropDownButton';

storiesOf('DropDownButton', module).add('basic example', () => (
  <DropDownButton onToggle={action('onToggle')} buttonLabel="Toggle button">
    {({ dropDownRef }) => <Window ref={dropDownRef}>DropDown content</Window>}
  </DropDownButton>
));
