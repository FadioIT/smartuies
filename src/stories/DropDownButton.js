import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Button, Input, Window } from '../utils/storyBookUtils';

import DropDownButton from '../components/DropDownButton';

storiesOf('DropDownButton', module)
  .add('basic example', () => (
    <DropDownButton
      onToggle={action('onToggle')}
      renderButton={({ buttonRef, onToggle, open }) => (
        <Button ref={buttonRef} onClick={onToggle}>
          {open ? 'Close dropDown' : 'Open dropDown'}
        </Button>
      )}
    >
      {({ dropDownRef }) => <Window ref={dropDownRef}>DropDown content</Window>}
    </DropDownButton>
  ))
  .add('input as button', () => (
    <DropDownButton
      onToggle={action('onToggle')}
      renderButton={({ buttonRef, onToggle, open }) => (
        <Input ref={buttonRef} onFocus={!open && onToggle} value="Focus me" />
      )}
    >
      {({ dropDownRef }) => <Window ref={dropDownRef}>DropDown content</Window>}
    </DropDownButton>
  ));
