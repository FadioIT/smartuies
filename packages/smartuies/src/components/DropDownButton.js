import React from 'react';
import PropTypes from 'prop-types';
import PopUpAnchor from './PopUpAnchor';

class DropDownButton extends React.Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    renderButton: PropTypes.func.isRequired,
    onToggle: PropTypes.func,
  };

  state = { open: false };

  dropDownRef = React.createRef();

  buttonRef = React.createRef();

  onToggle = () => {
    this.setState(
      ({ open }) => ({ open: !open }),
      () => {
        if (this.state.open) {
          this.addWindowEventHandlers();
        } else {
          this.removeWindowEventHandlers();
        }
        if (this.props.onToggle) {
          this.props.onToggle(this.state.open);
        }
      },
    );
  };

  onWindowMouseDown = event => {
    const dropdown = this.dropDownRef.current;
    const button = this.buttonRef.current;
    if (dropdown) {
      let { target } = event;
      while (target) {
        if (target === dropdown || target === button) {
          return;
        }
        target = target.parentNode;
      }
    }
    this.onToggle();
  };

  addWindowEventHandlers = () => {
    window.addEventListener('resize', this.onToggle);
    window.addEventListener('mousedown', this.onWindowMouseDown);
  };

  removeWindowEventHandlers = () => {
    window.removeEventListener('resize', this.onToggle);
    window.removeEventListener('mousedown', this.onWindowMouseDown);
  };

  componentWillUnmount() {
    this.removeWindowEventHandlers();
  }

  render() {
    const { children, renderButton, ...props } = this.props;
    const { open } = this.state;
    return (
      <PopUpAnchor
        {...props}
        displayPopUp={open}
        anchor={renderButton({
          ...props,
          buttonRef: this.buttonRef,
          onToggle: this.onToggle,
          open,
        })}
      >
        {children({
          dropDownRef: this.dropDownRef,
          onToggle: this.onToggle,
          open,
        })}
      </PopUpAnchor>
    );
  }
}

export default DropDownButton;
