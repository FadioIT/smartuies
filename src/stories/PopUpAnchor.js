import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';
import { Button, Window } from '../utils/storyBookUtils';

import PopUpAnchor, { PopUpPosition } from '../components/PopUpAnchor';

storiesOf('PopUpAnchor', module)
  .addDecorator(withKnobs)
  .add('basic example', () => {
    const popUpHeightMatchesAnchorHeight = boolean(
      'Match anchor height',
      false,
    );
    const popUpWidthMatchesAnchorWidth = boolean('Match anchor width', false);
    const displayPopUp = boolean('displayPopUp', true);
    const popUpPosition = select(
      'popUpPosition',
      {
        [PopUpPosition.LEFT]: 'Left',
        [PopUpPosition.RIGHT]: 'Right',
        [PopUpPosition.ABOVE]: 'Above',
        [PopUpPosition.BELOW]: 'Below',
        [PopUpPosition.TOP_LEFT]: 'Top left',
      },
      PopUpPosition.BELOW,
    );

    return (
      <PopUpAnchor
        popUpHeightMatchesAnchorHeight={popUpHeightMatchesAnchorHeight}
        popUpWidthMatchesAnchorWidth={popUpWidthMatchesAnchorWidth}
        popUpPosition={popUpPosition}
        anchor={<Button>Anchor element</Button>}
        displayPopUp={displayPopUp}
      >
        <Window>PopUp</Window>
      </PopUpAnchor>
    );
  });
