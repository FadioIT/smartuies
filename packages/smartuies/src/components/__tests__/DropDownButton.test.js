import React from 'react';
import { shallow, mount } from 'enzyme';
import DropDownButton from '../DropDownButton';
import PopUpAnchor from '../PopUpAnchor';

const getElement = () => (
  <DropDownButton
    renderButton={({ onToggle, open }) => (
      <button id="test-button" type="button" onClick={onToggle}>
        {open ? 'Close' : 'Open'}
      </button>
    )}
  >
    {({ dropDownRef }) => (
      <div id="test-container" ref={dropDownRef}>
        <div id="test-content">DropDown content</div>
      </div>
    )}
  </DropDownButton>
);

it('doesnâ€™t render dropdown when mounting', () => {
  const wrapper = shallow(getElement());
  expect(wrapper.find(PopUpAnchor).props().displayPopUp).toEqual(false);
  expect(wrapper).toMatchSnapshot();
});

it('opens the dropdown when calling onToggle', () => {
  const wrapper = mount(getElement());

  wrapper.find('#test-button').simulate('click');
  expect(wrapper.find(PopUpAnchor).props().displayPopUp).toEqual(true);
});
