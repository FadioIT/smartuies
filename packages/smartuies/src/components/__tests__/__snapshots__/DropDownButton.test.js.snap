// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesnâ€™t render dropdown when mounting 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DropDownButton
    renderButton={[Function]}
  >
    [Function]
  </DropDownButton>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "anchor": <button
        id="test-button"
        onClick={[Function]}
        type="button"
      >
        Open
      </button>,
      "children": <div
        id="test-container"
      >
        <div
          id="test-content"
        >
          DropDown content
        </div>
      </div>,
      "displayPopUp": false,
      "popUpHeightMatchesAnchorHeight": false,
      "popUpPosition": "below",
      "popUpWidthMatchesAnchorWidth": false,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          id="test-content"
        >
          DropDown content
        </div>,
        "id": "test-container",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "DropDown content",
          "id": "test-content",
        },
        "ref": null,
        "rendered": "DropDown content",
        "type": "div",
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "anchor": <button
          id="test-button"
          onClick={[Function]}
          type="button"
        >
          Open
        </button>,
        "children": <div
          id="test-container"
        >
          <div
            id="test-content"
          >
            DropDown content
          </div>
        </div>,
        "displayPopUp": false,
        "popUpHeightMatchesAnchorHeight": false,
        "popUpPosition": "below",
        "popUpWidthMatchesAnchorWidth": false,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            id="test-content"
          >
            DropDown content
          </div>,
          "id": "test-container",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "DropDown content",
            "id": "test-content",
          },
          "ref": null,
          "rendered": "DropDown content",
          "type": "div",
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
